tasks/component-usage-graph.js

```diff
@@ -0,0 +1,174 @@
+import fs from 'fs';
+import path from 'path';
+import chalk from 'chalk';
+import { DepGraph } from 'dependency-graph';
+import { init, parse } from 'es-module-lexer';
+import { getComponentFiles } from '../rollup/rollup-common.js';
+
+/**
+ *
+ * @returns {DepGraph} - returns a component `DepGraph`.
+ *
+ */
+export async function getComponentsGraph () {
+  const componentsGraph = new DepGraph();
+  const componentsFileList = getComponentFiles();
+  await init;
+
+  // Importer: One component that has one or more imports.
+  // Imported: One of the imports needed by the importer.
+  // e.g: cc-notice needs cc-icon. In this case, cc-notice is the importer and cc-icon the imported.
+
+  for (const importerFilePath of componentsFileList) {
+    const importerFile = fs.readFileSync(importerFilePath, 'utf8');
+    const [imports] = parse(importerFile);
+
+    for (const imported of imports) {
+
+      const importerRelative = path.relative(process.cwd(), importerFilePath);
+      const importerRelativeDir = path.dirname(importerRelative);
+      const sourceRelative = path.join(importerRelativeDir, imported.n);
+
+      if (!componentsGraph.hasNode(importerRelative)) {
+        componentsGraph.addNode(importerRelative);
+      }
+
+      const isImporterInList = componentsFileList.includes(importerRelative);
+      const isImportedInList = componentsFileList.includes(sourceRelative);
+      const isImportedLocal = imported.n.startsWith('./') || imported.n.startsWith('../');
+
+      if (isImporterInList && isImportedInList && isImportedLocal && !componentsGraph.hasNode(sourceRelative)) {
```

Comment:
**suggestion:** looks like `isImporterInList && isImportedInList && isImportedLocal` is common to both your conditions (l40 and l44), maybe we could move these into their own `if`?

I see two options but I might be wrong:

```js
if (isImporterInList && isImportedInList && isImportedLocal) {
  if (componentsGraph.hasNode(sourceRelative)) {
    componentsGraph.addNode(sourceRelative);
  }

  componentsGraph.addDependency(importerRelative, sourceRelative);
}
```
or (really not sure about this one, I never use `continue` :eyes:)
```js
if (!isImporterInList || !isImportedInList || !isImportedLocal) {
  continue;
}

if (!componentsGraph.hasNode(sourceRelative)) {
  componentsGraph.addNode(sourceRelative);
}

componentsGraph.addDependency(importerRelative, sourceRelative);
```
or just move `isImporterInList && isImportedInList && isImportedLocal` to a `const` (but I don't know how to name it) and reuse it:
```js
const areInListAndLocal = isImporterInList && isImportedInList && isImportedLocal;

if (areInlistAndLocal && !componentsGraph.hasNode(sourceRelative)) {
  componentsGraph.addNode(sourceRelative);
}

if (areInListAndLocal) {
  componentsGraph.addDependency(importerRelative, sourceRelative);
}
```

Reply:
On some previous iterations the option 1 was the one implemented. Though, I didn't stick with it as I don't like having multiple if nested if I can avoid them. However, I'll change it for option 3 if that's okay for you :wink:

Reply:
Yes no problem, even your version was ok for me :+1: